Mignoubou_DSR_Contribution

Mignoubou opened the pull request.

Mignoubou committed:
Added gas_king_rating in calculators

Mignoubou committed:
Forgot to remove personal comments

dima-stefantsov commented on calculators/gas_king_rating:
```diff
@@ -0,0 +1,123 @@
+<?php
+
+$GLOBALS['rating_calculators']['pink_color_rating'] = [
```
we already have a pink color rating :)

Mignoubou replied:
Fixed

dima-stefantsov commented on calculators/gas_king_rating:
```diff
@@ -0,0 +1,123 @@
+    'default_rating' => 2000,
+    'fields' => [ // always available: game_id, player_id, team, winner_team, status
+        'g.duration',
+        // 'g.sc2_version',
```
remove unused comments

Mignoubou replied:
Fixed

dima-stefantsov commented on calculators/gas_king_rating:
```diff
@@ -0,0 +1,123 @@
+
+$GLOBALS['rating_calculators']['pink_color_rating'] = [
+    'description' => '
+        gas king rating
```
first line of a custom rating should mention author, version. Like this
`by <a href="/player/125980">Memoria</a>: per-matchup calculation v1`

Mignoubou replied:
Should I simply put my name or do I have to respect this (html ?) notation ?
Because if I have to I do not know how to do it.

dima-stefantsov replied:
yes, you should give a link to yourself. You find your profile, get the number from address bar (your id), and you are done. See the memoria and dreamless ratings.
![image](https://github.com/dima-stefantsov/dsr_ratings/assets/1370909/99250c31-e36b-4006-9b11-93b77b2a7bb3)

Mignoubou replied:
Fixed ?

Mignoubou committed:
Removed useless fields from gas_king_rating

dima-stefantsov commented on calculators/gas_king_rating:
```diff
@@ -0,0 +1,123 @@
+            else
+            {
+                $min_timings = [1.5, 3, 4.5, 10];
+                $number_of_frames_in_a_min = 22.4 * 60;
```
you are about to create an array, and do this calculations more than 12 000 000 times in a row. Sometimes you gotta do what you've gotta do, but in this case you can put such constants right under line 49

Mignoubou replied:
Should be fixed everywhere

dima-stefantsov commented on calculators/gas_king_rating:
```diff
@@ -0,0 +1,123 @@
+                }
+            }
+
+            if (!($player['rating'] + $diff < 0))
```
is not !(<) same as >= ?

Mignoubou replied:
Yes it is, fixed

dima-stefantsov commented on calculators/gas_king_rating:
```diff
@@ -0,0 +1,101 @@
+    'main' => 'dsr_gas_king_rating__main',
+    ];
+
+function same_sign($a, $b)
```
name of every function you write must have same prefix

Mignoubou replied:
I think I fixed it, although I tried to understand what was a prefix but didn't quite understand

dima-stefantsov replied:
prefix is always `_name_of_rating__function_name()`, in your case it should be `_mignoubou_gas_king__main()`
in case of builtin ratings (not custom rating like the one you are making) I also prepend dsr_

Mignoubou committed:
Removed the assignement of variables outside of the loops

dima-stefantsov commented on calculators/gas_king_rating:
```diff
@@ -0,0 +1,101 @@
+            $diff = 0;
+            $duration_of_the_game_in_minutes = $teams[0][0]['duration'] / 22.4 / 60;
+
+            if($player['gas_count'] === 0)
```
does it work? I think field values such as this come in a string format, therefore you'll have to use == 0 comparison here

Mignoubou replied:
I honestly have no idea since this is my first code in php.

dima-stefantsov replied:
I know it does not work php-wise :) I'm asking that since that's the algorithm you've written, you must have noticed a part of your expected flow never happens. That's why I'm asking "do you notice this works?"

Mignoubou replied:
I tried with intval (on the team duration not on the entire fraction) and without it and it didn't change any rating from what I've seen

dima-stefantsov replied:
gas_count is actually a variable that is used for faster database queries. In code there's no real need to use it, you could use count($player['gas_timings']), but using it is fine too.

Mignoubou committed:
Yes it is.

"!(<)" = ">="

Mignoubou committed:
Changed === into == on line 38

Mignoubou committed:
Is this what you meant by prefix ? I have no idea of what it is (Line 21)

dima-stefantsov commented on calculators/gas_king_rating:
```diff
@@ -0,0 +1,101 @@
+    'main' => 'dsr_gas_king_rating__main',
+    ];
+
+function dsr_gas_king_rating__same_sign($a, $b)
```
right after the meta info, first function should be the main one. For the readability purposes. Only later, where noone reads, you declare helper functions. Recommendations are to place functions in order they are getting called. This is no C, functions work no matter where you've declared it.

Mignoubou replied:
Ok

Mignoubou replied:
I moved the sign function at the end of the code, fixed

Mignoubou committed:
Added author in the description

dima-stefantsov commented on calculators/gas_king_rating:
```diff
@@ -0,0 +1,101 @@
+
+                        else
+                        {
+                            $player['rating'] += intval($diff / 10);
```
$player[rating] must only have one assignment. That is what $diff used for. You are using it for different purposes :)

treat assigning player[rating] as `return`, because it essentially is

Mignoubou replied:
Fixed

Mignoubou committed:
Moved the same_sign function at the end of the program

Mignoubou committed:
Changed the assignement of rating and changed a bool into the same_sign function for readability

dima-stefantsov commented on calculators/gas_king_rating:
```diff
@@ -0,0 +1,101 @@
+<?php
+
+$GLOBALS['rating_calculators']['gas_king_rating'] = [
```
custom rating names start with _ (same for function prefixes which mimic rating name)

then follows the name of the author, then rating name

having "rating" in rating name is redundant

it should be something like "_pewpewpew_gas_king", or "_mingoubou_gas_king", or "_ppp_gasking", or "_noadkoko_gasking"...

Mignoubou replied:
Should I only change it on this line or should my prefix also not contain "rating" and contain "_" before it and my username ?

dima-stefantsov replied:
your prefix must be the same as your rating name. Refer to _memoria_matchups.

Mignoubou committed:
I think this fixes the prefix problem (need verification)

Mignoubou committed:
Changed the name of the rating to _mignoubou_gas_king

Mignoubou committed:
Added php extension

dima-stefantsov commented on calculators/gas_king_rating.php:
```diff
@@ -0,0 +1,103 @@
+        by <a href="/player/22267">PewPewPew</a>: gas king rating v1
+        only objective: gas as much and as soon as you can
+
+        you start with 2000 rating (compression is activated although maybe not efficient, I was too lazy to see how you did it)
```
this text is a description for the public page like https://ds-rating.com/rating/level 
![image](https://github.com/dima-stefantsov/dsr_ratings/assets/1370909/1d6c3495-989f-41ab-9bdb-95c615d3361a)

you probably don't want to announce how lazy you are, and definitely people reading will not understand what they did that you mention.

Mignoubou replied:
OH ok I didn't know, I'll definitely change this then yes.

Mignoubou replied:
People won't be blatantly aware that I'm lazy anymore, fixed

Mignoubou committed:
Moved the rating assignement in the necessary condition, changed the description

Mignoubou committed:
Removed a lot of useless intval functions for the same result

Although I have not managed to get integer values for the rating somehow, even if the only function that adds or removes rating is done with integers, so I am deeply confused.

Mignoubou committed:
Updated gas_king to v2

Mignoubou reviewed (commented):
Added a penality for losing while gassing intensively, didn't manage to fix the issue with floating values in the ratings (I will try again later), didn't change anything while winning.
I also need to change the description to fit the new version better and change something line 58.

Mignoubou commented:
> Added a penality for losing while gassing intensively, didn't manage to fix the issue with floating values in the ratings (I will try again later), didn't change anything while winning.
> I also need to change the description to fit the new version better and change something line 58.

Fix the rating assignement in case there is no gas by replacing by a diff

Mignoubou committed:
(Maybe) fixed the rounding issue (tired, need to check later if I didn't add issues while solving)

Mignoubou committed:
Moved the game_duration array out of the loops

Mignoubou committed:
Added a comment, improved description, fixed line 58

Mignoubou committed:
changed line 90 to remove a useless operation

Mignoubou committed:
Replaced some assignements with +=, /= etc and replaced player['rating'] - 2000 by dist_to_2000

Mignoubou committed:
Wrong comment, line is used to prevent an "undefined" error

Mignoubou committed:
The condition was indeed useless, removed it.

Mignoubou committed:
Made a small modification line 76, to stop giving new players less rating on their first game (and people with exactly 2000 rating in general)

Mignoubou committed:
Fixed line 103 where I forgot to take into account the dist_to_2000 when adding the diff

Mignoubou committed:
Modified description

dima-stefantsov commented on calculators/gas_king_rating.php:
```diff
@@ -0,0 +1,122 @@
+                    }
+                }
+
+                $player['rating'] += round($diff);
```
your diffs can be negative, right? Then it is on you to assure rating will never go below 0, or 1, or whatever positive number you want to have as lowest number.
Like this: `$player['rating'] = max(1, $player['rating'] + $diff);`

Since you've set the starting rating to 2000, lowest should probably be 1, not 0.

dima-stefantsov replied:
also refer to the manual for the functions you are using, https://www.php.net/manual/en/function.round.php , it says round() returns float. Rating must contain int. Do it `$player['rating'] = max(1, $player['rating'] + intval($diff));`

Mignoubou replied:
Replaced the round by intval, I will check the diff problem tomorrow

Mignoubou replied:
I checked the diff, it couldn't go in the negatives but could reach 0, now it can't anymore, fixed

dima-stefantsov commented on calculators/gas_king_rating.php:
```diff
@@ -0,0 +1,122 @@
+
+$GLOBALS['rating_calculators']['_mignoubou_gas_king'] = [
+    'description' => '
+        by <a href="/player/22267">PewPewPew</a>: gas king rating v2
```
add a newline right after this subtitle, like done in other ratings

Mignoubou replied:
Done

dima-stefantsov commented on calculators/gas_king_rating.php:
```diff
@@ -0,0 +1,122 @@
+        only objective: gas as much and as soon as you can, but do not lose !
+        you start with 2000 rating
+        you lose 10 rating per minute for each gas not taken (without compression) (only when winning)
+        you win 1, 2, 3, 4 rating per minute for gas 1, 2, 3, 4 (without compression)
```
it is not clear what compression is. You maybe want to describe it. Up to you.

Mignoubou replied:
done

Mignoubou committed:
Added a space in the description

Mignoubou committed:
Replaced round by intval to respect the integer type of the ratings

Mignoubou committed:
Changed description

Mignoubou committed:
Now rating can't reach 0 anymore

Mignoubou committed:
Forgot to add the floor on the inner diff adding.

Mignoubou committed:
Update gas_king_rating.php to final (I hope) version

Modified rewards, compression, fixed bugs, fixed maths, should work now

Mignoubou committed:
Can't go to 0 rating anymore
